Color of the Bottom Show Answer Changes based on the theme color

neeldoshii opened the pull request:
## Fixes
* Fixes #14209


## Approach
From the theme changed the color of the button also replaced the hardcoded color code to theme color code using `@color/theme_xxx` so in future if the color code of the theme if changes then it will also get changed dynamically.

`<solid android:color="@color/theme_black_primary" />`

## Testing Instructions
 1. Change the theme of the app from (Hamburger Menu ---> Settings ---> Appearance ---> Select Theme)
 2. Choose the theme (Light, Plain, Black, Dark)
 3. Go to home. 
 4. Open any card and then check the show answer button color matches the theme of the app.
 5. Repeat step 1 till you test all theme (Light, Plain, Black, Dark)
 

## How Has This Been Tested?
This has been tested on physical device(Realme 3i) and on emulator (Google Pixel 3)

## Screenshots

### Light Mode
<img src="https://github.com/ankidroid/Anki-Android/assets/60827173/e0c9d528-4673-41cb-906a-33e08aeb5dca" width="50%" height="50%">

### Dark Mode

<img src="https://github.com/ankidroid/Anki-Android/assets/60827173/1e95dab7-37ae-4936-a0ef-842695d0b2f2" width="50%" height="50%">

### Plain Mode

<img src="https://github.com/ankidroid/Anki-Android/assets/60827173/0f0ffb0f-6053-4ee6-9017-56a3a8b83133" width="50%" height="50%">

### Black Mode

<img src="https://github.com/ankidroid/Anki-Android/assets/60827173/84bbca6c-0f79-4bd3-86ce-ca668143ed47" width="50%" height="50%">





## Checklist

- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ ] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [X] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

neeldoshii commented:
Special Thanks to @SaketKaswa20 & @david-allison

Your messages in the github issues helped a lot to me in fixing the issue. Thank you Saket & David.

BrayanDSO commented:
Hi there. The issue asks to change the **navigation bar color** to match the Show answer button. Here what you are doing is changing the Show answer button color (not the navigation bar's) and matching it with the toolbar

neeldoshii commented:
Does the issue means changing this bottom navigation color? @BrayanDSO 

<img width="215" alt="image" src="https://github.com/ankidroid/Anki-Android/assets/60827173/12bcc98d-a8d4-4730-aa13-57a1a56ad8da">

neeldoshii commented:
> > Does the issue means changing this bottom navigation color? @BrayanDSO
> 
> Yes, that one

Ok, my bad I misunderstood the bottom nav and thought of submit button color. WIll fix it in the next commit.

neeldoshii commented:
In the following commit following changes are made. 

> Does the issue means changing this bottom navigation color? @BrayanDSO
> 
> <img alt="image" width="215" src="https://private-user-images.githubusercontent.com/60827173/293575634-12bcc98d-a8d4-4730-aa13-57a1a56ad8da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4MzYyNTYsIm5iZiI6MTcwNDgzNTk1NiwicGF0aCI6Ii82MDgyNzE3My8yOTM1NzU2MzQtMTJiY2M5OGQtYThkNC00NzMwLWFhMTMtNTdhMWE1NmFkOGRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA5VDIxMzIzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5ZjdmYzBhOGVlZjVlYjRlMWY4MjkyNmMyOTkxNmViNmYzYTdmZmUwOWYzZmFhYzEwOWJjYWRjMWU5YjY0OWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qXag-kn8vCuN0sx2bS3t0bQc-ji4DEG4Z_61G5ks9mc">

As per the required changes. I have made the changes which changes the color based on the theme of the app.

**App Theme**

<br>

| Black Mode | Dark Mode | Light Mode | Plain Mode |
|------------|-----------|------------|------------|
| ![Black_Mode](https://github.com/ankidroid/Anki-Android/assets/60827173/1a72c67b-8ba7-4258-b462-861f5b9c1f8e) | ![Dark_Mode](https://github.com/ankidroid/Anki-Android/assets/60827173/e3853c96-9722-4a7e-897c-a84bf4813492) | ![Light_mode](https://github.com/ankidroid/Anki-Android/assets/60827173/50139171-1e97-4cc3-93d6-8c4d792b4a35) | ![Plain_Mode](https://github.com/ankidroid/Anki-Android/assets/60827173/c1ee4391-24e3-44c6-82bc-7c869916f2f1) |


In previous commit I changed the color code of footer_button_all_black,plain,hard_dark, color code due to misunderstanding so in this reverted those code back to original color code which were there earlier as its irrevelant to the issue.


@BrayanDSO is this as per requirement? 
I also have a optimised approach the code for navigation bar, currently in the page_fragment we have changed the color of window navigation bar but if we just move this to ankiActivity and give the whole app bottom navigation color based on the theme of the app will be good as per UX. What do you think?

BrayanDSO commented:
I beg your pardon, but there was a mistake from my part. The issue asks to change the navigation bar color to match the `Show answer` bar, not the toolbar.

neeldoshii commented:
> I beg your pardon, but there was a mistake from my part. The issue asks to change the navigation bar color to match the `Show answer` bar, not the toolbar.

Can you show it with screenshot I am bit confused as thats what I tried to change it last time?

BrayanDSO commented:
The navigation bar (region I marked with a green circle) should have the same color of the show answer button (red square)

![image](https://github.com/ankidroid/Anki-Android/assets/69634269/987d94dd-e965-4591-9589-5d8aa807b6dc)

Don't change the show answer button color, only the navigation bar's

neeldoshii commented:
> The navigation bar (region I marked with a green circle) should have the same color of the show answer button (red square)
> 
> ![image](https://private-user-images.githubusercontent.com/69634269/295736283-987d94dd-e965-4591-9589-5d8aa807b6dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ5Mjc3MTUsIm5iZiI6MTcwNDkyNzQxNSwicGF0aCI6Ii82OTYzNDI2OS8yOTU3MzYyODMtOTg3ZDk0ZGQtZTk2NS00NTkxLTk1ODktNWQ4YWE4MDdiNmRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDIyNTY1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ5MzE2NjdmODg5NjgzYmI4ZTM5ZWYyNTBlMTc4ZDIyNjYwMDUwZDZiYTRjYzJmMGI3ZTgwMThmYmI0MTMwZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WyNFK4PD6JDl3f_hM0LIdbMyGaMFFx1RpvcDZ6Gfi9Y)
> 
> Don't change the show answer button color, only the navigation bar's

In the current commit I have changed it to the theme of the app as per requirements it should have the color of the show answer. Am I understanding right?

BrayanDSO commented:
Yes

neeldoshii commented:
In the following commit following changes are made. 

**App Theme**

<br>

| Black Mode | Dark Mode | Light Mode | Plain Mode |
|------------|-----------|------------|------------|
| ![Black_Mode](https://github.com/ankidroid/Anki-Android/assets/60827173/0bb5c24f-c475-45f0-a417-c9151eb1e929) | ![Dark_Mode](https://github.com/ankidroid/Anki-Android/assets/60827173/6a9f193e-c2ec-4974-9d66-9c397902daf4) | ![Light_mode](https://github.com/ankidroid/Anki-Android/assets/60827173/c2cf08ac-e37f-465e-890f-3a16e4af6bb4) | ![Plain_Mode](https://github.com/ankidroid/Anki-Android/assets/60827173/0977b3c8-9557-4a11-9544-44392fdc217a) |



@BrayanDSO Can you review this?

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.kt:
```diff
@@ -179,6 +179,28 @@ open class Reviewer :
         mTextBarReview = findViewById(R.id.review_number)
         mToolbar = findViewById(R.id.toolbar)
 
+        when (currentTheme.id) {
+            /*
+            Corresponding ID Reference to theme
+            //1 -> Theme_Light
+            //2--> Theme_Plain
+            //3--> Theme_Black
+            //4--> Theme_Dark
+             */
+
+            "1" ->
+                window!!.navigationBarColor =
+                    ContextCompat.getColor(applicationContext, R.color.material_blue_grey_700)
+            "2" ->
+                window!!.navigationBarColor =
+                    ContextCompat.getColor(applicationContext, R.color.theme_plain_primary)
+            "3" ->
+                window!!.navigationBarColor =
+                    ContextCompat.getColor(applicationContext, R.color.theme_black_primary)
+            "4" ->
+                window!!.navigationBarColor =
+                    ContextCompat.getColor(applicationContext, R.color.material_blue_grey_800)
+        }
```
Can be replaced with:

```kotlin
import androidx.appcompat.widget.ThemeUtils

// ...

window!!.navigationBarColor = ThemeUtils.getThemeAttrColor(this, android.R.attr.colorAccent)
```

neeldoshii replied:
`window!!.navigationBarColor = ThemeUtils.getThemeAttrColor(this, android.R.attr.colorAccent)`

Not sure how to match the same color of Show Answer as in attr.

david-allison replied:
With apologies, the above code was incorrect: we use `hardButtonRef`, which is not a color, so this will be more tricky

We can do better by extracting the color to a resource. I'll post sample code when I've worked through the issue

david-allison replied:
I looked into this futher:
* Our 'show answer' uses a drawable rather than a color
* It uses `hardButtonRef`, which is not a color (as it handles state changes)
* We can't easily change these colors to refer to an attribute
   * The 'Plain' theme uses one ref for all buttons: it wouldn't be correct to change this to point to an attr based on showAnswer

----

**Implementation**

The better way to handle this would be:

* define showAnswerColor in attrs.xml
* define showAnswerColor in theme_light/plain/dark/black.xml using the values you provided
* use `ThemeUtils.getThemeAttrColor(this, R.attr.showAnswerColor)`

neeldoshii replied:
I have made the appropriate changes 

```
            "1" ->
                window!!.navigationBarColor = ThemeUtils.getThemeAttrColor(this, R.attr.showAnswerColor)
            "2" ->
                window!!.navigationBarColor = ThemeUtils.getThemeAttrColor(this, R.attr.showAnswerColor)
            "3" ->
                window!!.navigationBarColor = ThemeUtils.getThemeAttrColor(this, R.attr.showAnswerColor)
            "4" ->
                window!!.navigationBarColor = ThemeUtils.getThemeAttrColor(this, R.attr.showAnswerColor)
            else -> {
                window!!.navigationBarColor = Color.TRANSPARENT
            }

```

The updated code now uses ThemeUtils and also added else block if any error occurs or by any chance in future a new theme is added we will have a transparent color rather than app leading to get crashed.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/pages/PageFragment.kt:
```diff
@@ -50,6 +50,7 @@ abstract class PageFragment : Fragment() {
             webViewClient = this@PageFragment.webViewClient
             webChromeClient = this@PageFragment.webChromeClient
         }
+
```
Could you revert this empty line change - no point in adding to the history of a file if it's unnecessary

david-allison reviewed (changes requested):
I remember searching for this for a while and not being able to find the right variable. Great job! Works functionally, just needs a little cleanup

For discussion: screenshot on API 23

<img width="356" alt="Screenshot 2024-01-11 at 11 49 04" src="https://github.com/ankidroid/Anki-Android/assets/62114487/4450bf50-90d7-41d7-adbf-1513ee770967">

david-allison reviewed (approved):
BEAUTIFUL! This is so clean, thank you!

david-allison commented:
@neeldoshii you can likely squash this and force push to remove the conflicts
